File: Data Structures & Algorithms/house-robber/submission-2.swift
class Solution {
    func rob(_ nums: [Int]) -> Int {
        var memo = Array(repeating: -1, count: nums.count)
        
        func dfs(_ i: Int) -> Int {
            guard i < nums.count else { return 0 }

            if memo[i] != -1 {
                return memo[i]
            }
            let skip = dfs(i+1)
            let rob = nums[i]+dfs(i+2)
            memo[i] = max(skip, rob)

            return memo[i]
        }

        return dfs(0)
    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1 :
            return nums[0]
        
        def robArr(arr: List[int]) -> int:
            visited = {}
            def dfs(i: int) -> int:
                if i >= len(arr): return 0

                if i in visited :
                    return visited[i]
                
                steal = arr[i]+dfs(i+2)
                skip = dfs(i+1)

                visited[i] = max(steal, skip)
                return max(steal, skip)
            return dfs(0)
        
        return max(
            robArr( nums[:-1] ), 
            robArr( nums[1:] )
            )
